lib/eslint/eslint-helpers.js

```diff
@@ -151,94 +285,176 @@ async function globSearch({ basePath, patterns, configs }) {
                 return false;
             }
 
-            const matchesPattern = matchers.some(matcher => matcher.match(relativePath));
+            /*
+             * Optimization: We need to track when patterns are left unmatched
+             * and so we use `unmatchedPatterns` to do that. There is a bit of
+             * complexity here because the same file can be matched by more than
+             * one pattern. So, when we start, we actually need to test every
+             * pattern against every file. Once we know there are no remaining
+             * unmatched patterns, then we can switch to just looking for the
+             * first matching pattern for improved speed.
+             */
+            const matchesPattern = unmatchedPatterns.size > 0
+                ? matchers.reduce((previousValue, matcher) => {
+                    const pathMatches = matcher.match(relativePath);
+
+                    // update unmatched patterns for potential errors
+                    if (pathMatches) {
+                        unmatchedPatterns.delete(matcher.pattern);
+                    }
+
+                    return pathMatches || previousValue;
+                }, false)
+                : matchers.some(matcher => matcher.match(relativePath));
 
             return matchesPattern && !configs.isFileIgnored(entry.path);
```

Comment:
Here, if the file is ignored, it still removes patterns from `unmatchedPatterns`, which seems wrong because when there are multiple patterns and at least one matches some non-ignored files, patterns that match only ignored files will not be reported.

For example, run:

```
eslint "lib/*.js" "tools/*.js" --ignore-pattern "tools/*.js"
```

In flat config mode, this command exits with 0 because `"lib/*.js"` matches something, although `"tools/*.js"` does not (it matches only ignored files).

In eslintrc mode, the same command exits with 1:

```
Oops! Something went wrong! :(

ESLint: 8.26.0

You are linting "tools/*.js", but all of the files matching the glob pattern "tools/*.js" are ignored.

...
```

I think that eslintrc behavior is correct and that flat eslint should work the same.

Reply:
Oh wow, nice catch. Yes this is a bug.